book/src/rfc/const_generic_fn.md

```diff
@@ -0,0 +1,224 @@
+## Summary
+This RFC is to support variable-length arguments and returned arrays in functions.
+
+## Motivation
+In the current implementation, array initialization requires explicitly stating all element values, which becomes impractical for large arrays. Furthermore, the lack of support for constant variables as array sizes limits the ability to write functions that generically handle arrays of varying sizes. This also necessitates hardcoding loop ranges, reducing flexibility.
+
+Allowing to define array sizes with constant variables would enhance function reusability and improve modularization.
+
+## Detailed design
+One of the useful features of const generic is to allow the declaration of default array or return array with const variable as size. Before making it possible to unlock these features, we need a way to propagate the actual values for the constant variables. This is crucial for doing sanity checks during circuit generation, such as checking the bounds of array access.
+
+To propagate the constant values, it needs to retain the structures for the variables from the bottom up. These retained structures would be also useful for builtin functions to do sanity checks on the inputs and outputs. For the support of const generic on types in the future, it would probably need to propagate the actual values in the same way.
+
+### Glossary
+- const: a keyword that can be used in different places. For example, when used in front of an argument name, in a function signature, it dictates that the function must be called with a value decided at compile time (for example, with a literal, or a global constant)"
+
+### Resolving const variables
+When the size of an array declaration is a literal number, the size can be resolved by calling the `size_of` on the array type. Then the `ExprKind::ArrayAccess` can check the bounds of the access by referencing the number returned from `size_of` call on an array type.
+
+When it is a constant variable as array size, we can't rely on the `size_of` to determine the size of the array directly through the type anymore, as the type checker can't resolve these values under its current design. For example, the type checker can't determine wether the length of the array matches with the return type:
+
+```rust
+fn thing(const len: Field) -> [Field; len] {
+  return [0; 3];
+}
+```
+
+That is because the type checker doesn't know the actual value of `len`, which needs to be propagated when computing the `Expr` during circuit generation.
+
+This RFC proposal is to defer the resolution of the actual values for the variable sizes of arrays to the circuit generation phase. The resolution can be done via propagating the actual sizes from the bottom up. This means to store the resolved values in a structured way, so they can be accessible in the computations of `compute_expr`, which is to compute the structured expressions. 
```

Comment:
`compute_expr` is not a type checker function, this is why I'm confused by this section

Reply:
IMO we should split the problem in two parts:

1. what does type checking needs to do? And what prevents it from doing its job with the current impl
2. what does circuit synthesizer needs to do? And what prevents it from doing its job with the current impl

I'm not sure I understand the answer to both question ATM.

Reply:
another question I have is, why not solve this by using two different `size_of` functions for type checking and circuit synthesizing? Or perhaps augment that function with a `ctx: Option<_>` argument which is `None` at type checking, but contains the actual instantiation at circuit synthesizing. Something like this perhaps:

```rust
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum TyKind {
    // TRUNCATED...
    /// An array of a fixed size.
    Array(Box<TyKind>, u32),
    /// An array of a generic size.
    GenericArray(Box<TyKind>, String),
    // TRUNCATED...
}

// we assume that Mul and Add are implemented for this type
enum TySize {
    Concrete(usize),
    Generic(String),
    Mul(TySize, TySize),
    Add(TySize, TySize),
}

impl<B: Backend> TypeChecker<B> {
    pub(crate) fn size_of_concrete(&self, ctx: HashMap<String, usize>, typ: &TyKind) -> usize {
        match self.size_of(Some(ctx), typ) {
            TySize::Concrete(size) => size,
            _ => panic!("this should not happen"),
        }
    }

    pub(crate) fn size_of(&self, ctx: Option<HashMap<String, usize>>, typ: &TyKind) -> TySize {
            match typ {
                // TRUNCATED...
                TyKind::GenericArray(typ, len_var) => {
                    if let Some(ctx) = ctx {
                        let size_of_typ = match self.size_of(ctx, typ) {
                            TySize::Concrete(size) => size,
                            _ => panic!("this should not happen"),
                        };
                        TySize::Concrete(ctx[len_var] * size_of_typ)
                    } else {
                        TySize::Mul(
                            self.size_of(None, typ),
                            TySize::Generic(len_var.clone()),
                        )
                    }
                }
            }
        }
```

Reply:
In terms of the use of `ctx`, I think here is a scenario it wouldn't work:

```rust
fn gen_arr(const size) -> [field; size] {
	return [0; size]; 
	// the FnEnv only store local context
	// so it either needs to:
	// 1. let VarOrRef carries the size value
	// 2. let FnEnv store additional meta info for VarOrRef, but the problem is FnEnv is designed to be scoped. It will need to propagate the meta info from this local scope to outer scopes, which is complicated and doesn't align with the purpose of FnEnv.
}

fn compare_arr(const size, a: [field; size], b: [field; size]) -> bool {
	...
}

let arr1 = gen_arr(10); // arr1 should carry the meta info for size

let arr2 = [0; 9];

// so the meta info can be used to check the arguments
assert_eq(compare_arr(arr1, arr2), true); 
```

I think we need to let the `VarOrRef` to carry the additional info, as these variables are passing between different scopes, among which it seems to be difficult to trace values directly via `FnEnv` which is scoped. But to achieve that with `VarOrRef`, `VarOrRef` has to be structured like `ComputedExpr`.

So I left this approach for now, and am investigating the other approaches.

Reply:
yeah, but it might work if we require the front-end syntax to enforce explicit declaration like
```rust
let arr1: [field; 10] = gen_arr(10);
```

I am not sure if this can complicate the dev experience. This requires the coders to do the calculation themselves, and this can be very cumbersome, especially const value returned from a generic function involves complicated calculations, such as 

```rust
fn gen_arr(const size) -> [field; size * 123] {
```

Reply:
my take is that monomorphization is probably the way, and if we need to force the dev to write the types to get that information we should do that (we can always implement inference later if we want it)